### PR TITLE
Domains: Create new "Add DNS record" page

### PR DESCRIPTION
**Pull request by leonardost**
### Changes proposed in this Pull Request

This PR creates a new "Add a new DNS record" page, accessible via the `/domains/manage/:domain/add-dns-record/:site` URL. It can also be accessed from the new "DNS records" page, via the "Add a new record" button in the breadcrumbs area. Originally, the add a new DNS record form was located in the same page where all DNS records for a domain were listed. This is part of the domain management pages redesign project detailed in pcYYhz-m2-p2.

#### Screenshots

This is the new page:

Screen Shot 2021-11-08 at 19 53 49

New page in mobile view:

Screen Shot 2021-11-08 at 20 06 55

This is where this functionality was located before:

Annotation on 2021-11-08 at 20-01-02

### Testing instructions

Access the live Calypso link, navigate to "Upgrades > Domains > Change your name servers and DNS records > DNS records", append the `?flags=domains/dns-records-redesign` flag to the URL and click on the "Add a new record" button. Alternatively, access the `/domains/manage/:domain/add-dns-record/:site` URL directly (the feature flag is not needed there because it's a new path).

Try adding some DNS records and ensure they are created correctly. Please also check the mobile view.

### CODE REVIEW

File: client/my-sites/domains/domain-management/add-dns-record/a-record.jsx [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+	};
+
+	render() {
+		const { fieldValues, isValid, onChange, selectedDomainName, show, translate } = this.props;
+		const classes = classnames( { 'is-hidden': ! show } );
+		const isNameValid = isValid( 'name' );
+		const isDataValid = isValid( 'data' );
+		const isAaaaRecord = fieldValues.type === 'AAAA';
+
+		let namePlaceholder = translate( 'Enter subdomain', {
```
Comment (a8ci18n) [inline client/my-sites/domains/domain-management/add-dns-record/a-record.jsx]: :information_source: String reuse speeds up translation and improves consistency. The following string might make a good alternative and has already been translated **25** times:
`translate( 'Enter subdomain (optional)', { context: 'Placeholder shown when entering the optional subdomain part of a new DNS record'} )` *ES Score: 10*
See 2 additional suggestions in the [PR translation status page](https://translate.wordpress.com/localci/status/automattic/wp-calypso/add/add-dns-record-page)


:information_source: This **context** is really long. Are you sure you don't want to use a translator comment instead?

File: client/my-sites/domains/domain-management/add-dns-record/a-record.jsx [hunk 2]
```diff
@@ -0,0 +1,77 @@
...
+				args: {
+					example: '2001:500:84::b',
+				},
+			} );
+		}
+
+		return (
+			<div className={ classes }>
+				<FormFieldset>
+					<FormLabel>{ translate( 'Name (optional)', { context: 'DNS Record' } ) }</FormLabel>
```
Comment (a8ci18n) [inline client/my-sites/domains/domain-management/add-dns-record/a-record.jsx]: :information_source: This string doesn't exist elsewhere without a **context** (or with a different one). You can probably drop the context. If needed, you can use a translator comment instead to clarify meaning.

File: client/my-sites/domains/domain-management/dns/a-record.jsx [hunk 3]
```diff
@@ -23,19 +23,20 @@ class ARecord extends Component {
 		const isDataValid = isValid( 'data' );
 		const isAaaaRecord = fieldValues.type === 'AAAA';
 
-		let namePlaceholder = translate( 'Enter subdomain (optional)', {
-			context: 'Placeholder shown when entering the optional subdomain part of a new DNS record',
+		let label = translate( 'Name (optional)', { context: 'DNS record' } );
```
Comment (a8ci18n) [inline client/my-sites/domains/domain-management/dns/a-record.jsx:26]: :information_source: This string doesn't exist elsewhere without a **context** (or with a different one). You can probably drop the context. If needed, you can use a translator comment instead to clarify meaning.

File: client/my-sites/domains/domain-management/dns/a-record.jsx [hunk 4]
```diff
@@ -23,19 +23,20 @@ class ARecord extends Component {
...
+		if ( isAaaaRecord ) {
+			label = translate( 'Name (required)', { context: 'DNS record' } );
```
Comment (a8ci18n) [inline client/my-sites/domains/domain-management/dns/a-record.jsx]: :information_source: String reuse speeds up translation and improves consistency. The following string might make a good alternative and has already been translated **89** times:
`translate( 'Name (required)' )` *ES Score: 11*
See 2 additional suggestions in the [PR translation status page](https://translate.wordpress.com/localci/status/automattic/wp-calypso/add/add-dns-record-page)


:information_source: This string already exists without a **context**. Only add a context if the meaning of the string is very specific.

File: client/my-sites/domains/domain-management/dns/a-record.jsx [hunk 5]
```diff
@@ -23,19 +23,20 @@ class ARecord extends Component {
...
+		}
+
+		const namePlaceholder = translate( 'Enter subdomain', {
```
Comment (a8ci18n) [inline client/my-sites/domains/domain-management/dns/a-record.jsx:31]: :information_source: String reuse speeds up translation and improves consistency. The following string might make a good alternative and has already been translated **16** times:
`translate( 'Enter subdomain (optional)', { context: 'Placeholder shown when entering the required subdomain part of a new DNS record'} )` *ES Score: 10*
See 2 additional suggestions in the [PR translation status page](https://translate.wordpress.com/localci/status/automattic/wp-calypso/add/add-dns-record-page)


:information_source: This **context** is really long. Are you sure you don't want to use a translator comment instead?

File: client/my-sites/domains/domain-management/dns/a-record.jsx [hunk 6]
```diff
@@ -23,19 +23,20 @@ class ARecord extends Component {
 		const isDataValid = isValid( 'data' );
 		const isAaaaRecord = fieldValues.type === 'AAAA';
 
-		let namePlaceholder = translate( 'Enter subdomain (optional)', {
-			context: 'Placeholder shown when entering the optional subdomain part of a new DNS record',
+		let label = translate( 'Name (optional)', { context: 'DNS record' } );
+		if ( isAaaaRecord ) {
+			label = translate( 'Name', { context: 'DNS record' } );
```
Comment (a8ci18n) [inline client/my-sites/domains/domain-management/dns/a-record.jsx:28]: :information_source: String reuse speeds up translation and improves consistency. The following string might make a good alternative and has already been translated **36** times:
`translate( 'Name', { context: 'Dns Record'} )` *ES Score: 13*
See 2 additional suggestions in the [PR translation status page](https://translate.wordpress.com/localci/status/automattic/wp-calypso/add/add-dns-record-page)


:information_source: This string already exists with the following **contexts**:
- `null` (no context)
- `Dns Record`
- `Label for HTML form "Name" field in contact form builder`
- `link name`
- `meta name`
- `Name of link anchor (TinyMCE)`
- `term name`
- `theme name`

 Would it make sense to reuse one of the above?

File: client/my-sites/domains/domain-management/dns/txt-record.jsx [hunk 7]
```diff
@@ -40,10 +40,10 @@ class TxtRecord extends Component {
 		return (
 			<div className={ classes }>
 				<FormFieldset>
-					<FormLabel>{ translate( 'Name', { context: 'Dns Record' } ) }</FormLabel>
+					<FormLabel>{ translate( 'Name (optional)', { context: 'Dns Record' } ) }</FormLabel>
```
Comment (a8ci18n) [inline client/my-sites/domains/domain-management/dns/txt-record.jsx:43]: :information_source: This string doesn't exist elsewhere without a **context** (or with a different one). You can probably drop the context. If needed, you can use a translator comment instead to clarify meaning.

File: client/my-sites/domains/domain-management/dns/txt-record.jsx [hunk 8]
```diff
@@ -40,10 +40,10 @@ class TxtRecord extends Component {
...
 					<FormTextInputWithAffixes
 						name="name"
-						placeholder={ translate( 'Enter subdomain (optional)', {
+						placeholder={ translate( 'Enter subdomain', {
```
Comment (a8ci18n) [inline client/my-sites/domains/domain-management/dns/txt-record.jsx:46]: :warning: This change will be queued for retranslation. We'll display the English text until then.
:information_source: String reuse speeds up translation and improves consistency. The following string might make a good alternative and has already been translated **16** times:
`translate( 'Enter subdomain (optional)', { context: 'Placeholder shown when entering the required subdomain part of a new DNS record'} )` *ES Score: 10*
See 1 additional suggestions in the [PR translation status page](https://translate.wordpress.com/localci/status/automattic/wp-calypso/add/add-dns-record-page)

File: client/my-sites/domains/domain-management/dns/add-dns-record.jsx [hunk 9]
```diff
@@ -0,0 +1,202 @@
...
+		);
+
+		return (
+			<Main wideLayout className="add-dns-record">
+				<BodySectionCssClass bodyClass={ [ 'edit__body-white' ] } />
+				<div className="add-dns-record__fullwidth">
+					{ this.renderBreadcrumbs() }
+					<FormattedHeader
+						brandFont
+						headerText={ translate( 'Add a new DNS record' ) }
```
Comment (a8ci18n) [inline client/my-sites/domains/domain-management/dns/add-dns-record.jsx:108]: :information_source: String reuse speeds up translation and improves consistency. The following string might make a good alternative and has already been translated **24** times:
`translate( 'Add new DNS record' )` *ES Score: 7*
See 1 additional suggestions in the [PR translation status page](https://translate.wordpress.com/localci/status/automattic/wp-calypso/add/add-dns-record-page)

File: client/my-sites/domains/domain-management/dns/add-dns-record.jsx [hunk 10]
```diff
@@ -0,0 +1,160 @@
...
+						isSubmittingForm={ dns.isSubmittingForm }
+						selectedDomainName={ selectedDomainName }
+						goBack={ this.goBack }
+					/>
+				</div>
+				<div className="add-dns-record__sidebar">
+					<div>
+						<strong>
+							{ translate( 'What are these used for?', {
+								comment: '"These" refers to DNS records',
```
Comment (delputnam) [inline client/my-sites/domains/domain-management/dns/add-dns-record.jsx:124]: 👍

**Review by delputnam — APPROVED** [review]
This LGTM. 👍 

We also need to have a way to "edit" records (which we don't currently have a way to do).

I'm thinking that we can do something like pass the record ID in a query string and then load the contents of that record into the fields. We'll also have to handle saving the record a little differently since there's no such thing as "editing" a record.

One other usability thing that has always bothered me is that we have always cleared the fields when there is an error, but I am thinking that perhaps we should leave them filled so that the user can edit the record info and try again  without re-typing everything. ...this feels somewhat more important now that we're going to allow "editing" of records.

This is already a pretty large PR, so maybe let's handle these as part of the PR to integrate this new page with the new DNS Records page. (Or even another task.)

Nice job on this one, as usual!

**Comment by leonardost** [comment]
Thanks for the review, @delputnam!
I've created a new task for editing DNS records here `1146722293412575-as-1201352849238785`, so I'll deploy this PR shortly.

**Comment by a8ci18n** [comment]
This Pull Request is now available for translation here: https://translate.wordpress.com/deliverables/6957523

Thank you @leonardost for including a screenshot in the description! This is really helpful for our translators.

**Comment by a8ci18n** [comment]
Translation for this Pull Request [has now been finished](https://translate.wordpress.com/deliverables/overview/6957523).

**Pull request merged**